File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        prefix = []
        postfix = [1 for i in range(len(nums))]
        i = 0

        for num in nums:
            if i == 0:
                prefix.append(num)
                i += 1
            else:
                prefix.append(num*prefix[i-1])
                i += 1
        for k in range(len(nums)-1, -1, -1):
            if k == (len(nums)-1):
                postfix[k] = nums[k]
            else:
                postfix[k] = postfix[k+1] * nums[k]

        result = []
        for i in range(len(nums)):
            if i == 0:
                result.append(postfix[i+1])
            elif i == (len(nums) - 1):
                result.append(prefix[i-1])
            else:
                result.append(prefix[i-1] * postfix[i+1])
        return result

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {

  let prefix = [nums[0]];
  nums.forEach((num, i) =>
    i !== 0 ? prefix.push(prefix[prefix.length - 1] * num) : ""
  );

  let postfix = [nums[nums.length - 1]];
  for (let i = nums.length - 2; i >= 0; i--) {
    postfix.unshift(nums[i] * postfix[0]);
  }

  const res = [];
  for (let i = 0; i < nums.length; i++) {
    if (i === 0) res.push(postfix[1]);
    else if (i === nums.length - 1) res.push(prefix[i - 1]);
    else res.push(prefix[i - 1] * postfix[i + 1]);
  }
  return res;
};